ecobee/auth.go

```diff
@@ -204,8 +232,38 @@ type Client struct {
 // (Application Key).  Use the Ecobee Developer Portal to create the
 // Application Key.
 // (https://www.ecobee.com/consumerportal/index.html#/dev)
+// This function is deprecated - use New() instead.
 func NewClient(clientID, authCache string) *Client {
-	return &Client{oauth2.NewClient(
-		context.Background(),
-		TokenSource(clientID, authCache))}
+	c, err := New(context.TODO(), clientID, Options{
+		AuthCache: fileAuthCache{authCache},
+	})
+
+	if err != nil {
+		// can't fail here.
+		panic("unexpected usage: " + err.Error())
+	}
+
+	return c
+}
+
+// Options specifies EcoBee client parameters.
+type Options struct {
+	ApplicationID string    // application ID created in Ecobee Developer Portal
+	AuthCache     AuthCache // token cache, typically TokenCacheFile(fileName)
+}
+
+// New creates a Ecobee API client for the specific options
+// Use the Ecobee Developer Portal to create the Application Key.
+// (https://www.ecobee.com/consumerportal/index.html#/dev)
+func New(ctx context.Context, clientID string, opt Options) (*Client, error) {
+	if opt.ApplicationID == "" {
```

Comment:
This is never used.  Was this supposed to check clientID?  I think I was mixing OAuth terminology with ecobee terminology.

Also, this will fail whenever NewClient is used, because it doesn't set it.

Looking at the ecobee portal, it's now just called an API key.     Since this is always required, I think it can be a normal parameter.

Reply:
Both parameters in Options are currently required, I think it's cleaner to pass them all through Options (except ctx which would look weird in a struct).

Removed clientID parameter here, renamed to appKey everywhere.

Reply:
AuthCache doesn't actually need to be required, although it would be kind of weird to not provide it.

I've been staring at the oauth2 library and wondering if we've built the right interface.     (Since it's hard to change later.)

https://github.com/golang/oauth2/issues/84 has given me some ideas, but I can't just copy that code.  

I may be overthinking this.

Reply:
Ok, I did some experimentation and it turns out most of the `ecobee.tokenSource` functionality can be replaced with `golang.org/x/oauth2` package which also makes it easy to get OAuth2 HTTP client.

Given that, if we added `ecobeekeeper.New(hc *http.Client)`, I can make OAuth2 work externally with relatively little effort:

```golang
cfg := &oauth2.Config{
  Endpoint: oauth2.Endpoint{
    TokenURL: "https://api.ecobee.com/token?client_id=3KPS9bjYiH....0MIZR",
  },
}

// load token from DB once at startup
var tok *oauth2.Token = loadTokenFromDB(db)
var ts *oauth2.TokenSource = cfg.TokenSource(ctx, tok)
// wrapper around TokenSource that saves tokens in DB whenever they change
ts = cachingTokenSource(ts, db) 
cli := ecobee.New(oauth2.NewClient(ctx, ts))
```

What do you think?

Internally we can build caching wrapper around `TokenSource` that writes to a file to support legacy API.

Also, we'd need to do something about intiial token fetching, but I think a simple command line tool like https://gist.github.com/jkowalski/20736a32a39ba85d9f7e630b6125e2f6 should do

Reply:
Hi Jarek,
  Sorry for not making any progress on this myself, I've been getting home from work and not wanting to spend more time with a computer.

  I like this direction.  It gets rid of a bunch of code and simplifies things.

  Do you need to put the client_id in the TokenURL, or can you just put it in the oauth2.Config directly?  

  I think we can re-use the TokenSource interface for a lot of the wrappers and it will be pretty clean.  Some of the examples in the bug linked before do something similar.

  Instead of the initial fetching that's in a separate command line tool, it could be a library so that CLI tools could just use it.  Maybe it could be a TokenSource?  

  One thing we don't support right now is the web based oauth flow (which is supported, you configure it in the ecobee portal.)  

  Instead of 100% backwards compatibility, it might be time to declare a v2 API and start using go mod and semantic versioning.  https://github.com/golang/go/wiki/Modules.  I'm open to that.

Thanks!

Reply:
Sounds like great ideas. I'm going to close this PR. I technically was able to use unmodified code + the tool from my GIST to load/store token in etcd and it works great. I'd love to collaborate on v2 efforts and port my usage to v2 when ready.